java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -370,177 +363,226 @@ public final void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+      currentStates.states.clear();
+    }
+
+    @Override
+    public void visitCaseGroup(CaseGroupTree tree) {
+      scan(tree.body());
     }
 
     @Override
     public void visitContinueStatement(ContinueStatementTree tree) {
-      currentStates = PackedStatementStates.instantiate();
+      currentStates = new PackedStatementStates(currentStates);
     }
 
     @Override
     public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
       Set<Symbol.VariableSymbol> assignedSymbols = extractor.findAssignedVariables(tree);
       invalidateAssignedVariables(assignedSymbols);
-      currentStates = evaluateStatement(currentStates, tree.statement());
-      PackedStatementStates nextStates = PackedStatementStates.instantiate();
-      for (ExecutionState state : currentStates) {
-        if (evaluateExpression(state, tree.condition()) != SymbolicBooleanConstraint.TRUE) {
-          nextStates.addState(state);
+      scan(tree.statement());
+      scan(tree.condition());
+      for (ExecutionState executionState : currentStates) {
+        if (executionState.popLast() != SymbolicValue.BOOLEAN_TRUE) {
+          nextStates.addState(executionState);
         }
       }
+      nextStates.addStates(currentStates.breakStates);
       currentStates = nextStates;
       invalidateAssignedVariables(assignedSymbols);
     }
 
     @Override
     public void visitExpressionStatement(ExpressionStatementTree tree) {
-      for (ExecutionState state : currentStates) {
-        evaluateExpression(state, tree.expression());
+      scan(tree.expression());
+      for (ExecutionState executionState : currentStates) {
+        executionState.popLast();
       }
     }
 
     @Override
     public void visitForStatement(ForStatementTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+      scan(tree.initializer());
       Set<Symbol.VariableSymbol> assignedSymbols = extractor.findAssignedVariables(tree);
       invalidateAssignedVariables(assignedSymbols);
       if (tree.condition() != null) {
-        PackedStatementStates nextStates = PackedStatementStates.instantiate();
-        for (ExecutionState state : currentStates) {
-          PackedStates conditionStates = evaluateCondition(state, tree.condition());
-          PackedStatementStates loopStates = evaluateStatement(conditionStates.trueStates, tree.statement());
-          if (!conditionStates.falseStates.isEmpty() || !loopStates.isEmpty()) {
-            state.mergeRelations(Iterables.concat(conditionStates.falseStates, loopStates));
-            nextStates.addState(state);
+        for (ExecutionState executionState : currentStates) {
+          for (ExecutionState conditionState : evaluateCondition(executionState, tree.condition())) {
+            SymbolicValue condition = conditionState.popLast();
+            if (condition == SymbolicValue.BOOLEAN_TRUE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)